cirq/protocols/json_serialization.py

```diff
@@ -623,23 +622,23 @@ def to_json(
         class ContextualEncoder(cls):  # type: ignore
             """An encoder with a context map for concise serialization."""
 
-            # This map is populated gradually during serialization. An object
-            # with components defined in this map will represent those
+            # These lists populate gradually during serialization. An object
+            # with components defined in 'context' will represent those
             # components using their keys instead of inline definition.
-            context_map: Dict[str, 'SerializableByKey'] = {}
+            context: List['SerializableByKey'] = []
+            keys: List[str] = []
 
             def default(self, o):
-                skey = getattr(o, '_serialization_key_', lambda: None)()
-                if skey in ContextualEncoder.context_map:
-                    if ContextualEncoder.context_map[skey] == o._json_dict_():
-                        return _SerializedKey(o)._json_dict_()
-                    raise ValueError(
-                        'Found different objects with the same serialization key:'
-                        f'\n{ContextualEncoder.context_map[skey]}\n{o}'
-                    )
-                if skey is not None:
-                    ContextualEncoder.context_map[skey] = o._json_dict_()
-                return super().default(o)
+                if not hasattr(o, '_serialization_name_'):
+                    return super().default(o)
+                try:
+                    idx = ContextualEncoder.context.index(o)
+                    return _SerializedKey(ContextualEncoder.keys[idx])._json_dict_()
+                except ValueError:
+                    ContextualEncoder.context.append(o)
+                    key = f'{o._serialization_name_()}_{len(ContextualEncoder.context)}'
+                    ContextualEncoder.keys.append(key)
+                    return super().default(o)
```

Comment:
here's our hacked version of the Encoder:

```
        obj = _ContextualSerialization(obj)

        class ContextualEncoder(cls):  # type: ignore
            """An encoder with a context map for concise serialization."""

            # These lists populate gradually during serialization. An object
            # with components defined in 'context' will represent those
            # components using their keys instead of inline definition.
            seen: Set[str] = set()

            def default(self, o):
                if not hasattr(o, '_serialization_name_'):
                    return super().default(o)
                for candidate in obj.object_dag[:-1]:
                    if candidate.obj == o:
                        if not candidate.key in ContextualEncoder.seen:
                            ContextualEncoder.seen.add(candidate.key)
                            return candidate.obj._json_dict_()
                        else:
                            return _SerializedKey(candidate.key)._json_dict_()
                raise ValueError("this should never happen")
```

Reply:
Namely, this removes the duplication of the suffix calculation and context building

Reply:
Added. Thanks for your help here!